### PR TITLE
Skip binaries in source tree rather than failing

### PR DESCRIPTION
**Pull request by haxtibal**
We're expecting source files to be UTF-8 text. Files that are invalid UTF-8 (most likely binary files) would have crashed StrictDoc. We can instead just skip them. Whatever they are, they are not traceable.

It's as simple as catching the UnicodeDecodeError. The majority of the
diff is just about remembering whether a file was successfully parsed
and indexed. Successfully parsed files are now in FileTraceabilityIndex

```
TracabilityIndex
  has 1 FileTraceabilityIndex
    has n SourceFileTraceabilityInfo
      has 1 SourceFile  # new: remember parsed/indexed source file
```

while any found file is (as before) in DocumentTree

```
Traceabilityindex
  has 1 DocumentTree
    has 1 SourceTree
      has n SourceFile  # remember *any* found source file
```

The binary files for testing was generated with
```
echo "00 08 00 00" | xxd -r -p > invalid_file.bin
```

### CODE REVIEW

File: tests/integration/features/file_traceability/20_skip_invalid_utf8/test.itest [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# file.bin was created with echo "00 08 00 00" | xxd -r -p > file.bin
+RUN: %strictdoc export %S --output-dir Output | filecheck %s --dump-input=fail
+CHECK: info: Skip tracing binary file
```
Comment (haxtibal) [inline tests/integration/features/file_traceability/20_skip_invalid_utf8/test.itest:3]: There are 3 such lines expected, one for each invalid input file. I'd like to handle them more specific, but couldn't figure out yet how to use filecheck with either regex or directory substitution. Will try to figure out later - hints are also appreciated :)

Comment (haxtibal) [reply tests/integration/features/file_traceability/20_skip_invalid_utf8/test.itest]: Ignore, nothing particular here. Just a typo on my side.

**Review by stanislaw — CHANGES_REQUESTED** [review]
Hi @haxtibal, thanks for taking care of this issue.

I feel like the PR is doing a bit more than just addressing the runtime error thrown by the binary files.

I like the idea of connecting Source File with Source File Traceabiity Info but I think the relation between these classes should be from SFTI to SF, not the other way round like you did. I explain my reasoning in the comments.

I would be fine with either seeing a reduced PR that only addresses the reading of binary files or also a refactoring PR but then let's invert the relation order.

Note that in general, I am have been struggling with maintaining a growing number of entiries and relations between them using just Python dictionaries. Everything is fine with creating dictionaries but when their number grows, it becomes harder to manage what needs to be tracked together when things get added/deleted/updated.

What you see in the traceability index as a 'graph database' was my first attempt to bring some order into how things could be stored in an in-memory graph and all contributing links would be maintained with generalized one-to-one or one-to-many relations. I was planning of moving to this solution as an intermediate step for `file_traceability_index` as well or maybe even merging it to the traceability index by adding more of the enum type relations to the enum below. Your attempt to link the source file with a source file info is a step in a similar direction but it does not address this higher-level issue that my brain is lacking some additional order in how this code is maintained. Hopefully this longer speech explaining the bigger picture in my head, helps.


```
class GraphLinkType(IntEnum):
    MID_TO_NODE = 1
    UID_TO_NODE = 2
    UID_TO_REQUIREMENT_CONNECTIONS = 3
    NODE_TO_INCOMING_LINKS = 4
    DOCUMENT_TO_PARENT_DOCUMENTS = 5
    DOCUMENT_TO_CHILD_DOCUMENTS = 6
    DOCUMENT_TO_TAGS = 7
```

Thanks!

**Comment by haxtibal** [comment]
> I like the idea of connecting Source File with Source File Traceabiity Info but I think the relation between these classes should be from SFTI to SF, not the other way round like you did. I explain my reasoning in the comments.

Ack. Should be doable as you lined it out here (thanks!). I first started the other way round because I found the existing but afaikt unused [SourceFile.traceability_info](https://github.com/strictdoc-project/strictdoc/blob/b59ab5f3829efcf61734f912af2f8b7fcef97e1a/strictdoc/core/finders/source_files_finder.py#L94) attribute and started using it without a look at object hierarchy, which would have made it clear.

>  or maybe even merging it to the traceability index by adding more of the enum type relations to the enum below

I'm afraid this could be a bit over my current time budget. Would you want me to do that? Otherwise I start with the less demanding SFTI -> SF approach.

**Comment by stanislaw** [comment]
> > I like the idea of connecting Source File with Source File Traceabiity Info but I think the relation between these classes should be from SFTI to SF, not the other way round like you did. I explain my reasoning in the comments.
> 
> Ack. Should be doable as you lined it out here (thanks!). I first started the other way round because I found the existing but afaikt unused [SourceFile.traceability_info](https://github.com/strictdoc-project/strictdoc/blob/b59ab5f3829efcf61734f912af2f8b7fcef97e1a/strictdoc/core/finders/source_files_finder.py#L94) attribute and started using it without a look at object hierarchy, which would have made it clear.

No worries. These things can be not obvious at all, and we don't talk through them upfront. It sounds like you got the idea 👍 

> 
> > or maybe even merging it to the traceability index by adding more of the enum type relations to the enum below
> 
> I'm afraid this could be a bit over my current time budget. Would you want me to do that? Otherwise I start with the less demanding SFTI -> SF approach.

No worries, I just mentioned the "graph db" approach to give you some context. You don't have to do that. Please proceed with only the SFTI-SF approach.

P.S.

By the way, I was thinking of doing 'strictdoc office hours'. It looks like with Discord I can implement this idea using a public video/audio/screen channel where everyone can join. I am thinking of doing every Wednesday 17-19:00 CET time. By default, I will just work on something strictdoc* during this time, and if someone would join, we could simply chat freely about whatever comes up.

I will not start doing it this week because the work is a little demanding currently, but very likely starting from next one. I also need to spend some time to document the office hours in the docs and send some announcements to Discord and dozen of StrictDoc users who only use email.

File: strictdoc/core/file_traceability_index.py [hunk 2]
```diff
@@ -420,16 +421,17 @@ def create_requirement(self, requirement: SDocNode) -> None:
...
     def create_traceability_info(
         self,
-        source_file_rel_path: str,
-        traceability_info: SourceFileTraceabilityInfo,
+        source_file: SourceFile,
         traceability_index,
     ) -> None:
-        assert isinstance(traceability_info, SourceFileTraceabilityInfo)
+        assert isinstance(
+            source_file.traceability_info, SourceFileTraceabilityInfo
```
[the remark this answers is not in the document]
Comment (stanislaw) [reply strictdoc/core/file_traceability_index.py]: > assert isinstance(source_file.traceability_info, SourceFileTraceabilityInfo)

Yes, sorry. Scratch this particular line from the consideration but the general comment still applies. To be honest I have been and being very short on time due to some overtime at work, so doing everything StrictDoc* in a rush. Wish to come back to more quiet days.

[the remark this answers is not in the document]
Comment (haxtibal) [reply strictdoc/core/file_traceability_index.py]: > the general comment still applies

Found and fixed one occurrence in the diff.

> I have been and being very short on time due to some overtime at work, so doing everything StrictDoc* in a rush. Wish to come back to more quiet days.

Please don't feel pushed by anything StrictDoc* and take quite days if possible. What you've already provided is awesome in its current state. This is not only my opinion, but feedback we get repeatedly from internal and external. It's way more then one could expect from a spare time project.

**Review by stanislaw — APPROVED** [review]
LGTM

Sorry for reviewing it much later. I had some extreme overtimes at work, and now it is time to relax.

**Pull request merged**